src/utils/label.ts

```diff
@@ -0,0 +1,56 @@
+import { closestElementCrossShadowBoundary, queryElementRoots } from "./dom";
+
+export interface CalciteLabelableComponent {
+  /**
+   * The host element.
+   */
+  readonly el: HTMLElement;
+
+  /**
+   * Text label.
+   */
+  label?: string;
+
+  /**
+   * The label this component is associated with.
+   */
+  labelEl: HTMLCalciteLabelElement;
+
+  /**
+   * Hook for components to provide custom label click behavior.
+   */
+  onLabelClick: (event: CustomEvent<any>) => void;
+}
+
+const labelTagName = "calcite-label";
+const labelClickEvent = "calciteInternalLabelClick";
+
+const findLabelForComponent = (componentEl: HTMLElement): HTMLCalciteLabelElement => {
+  const id = componentEl.id;
+  return (
+    (id && (queryElementRoots(componentEl, `${labelTagName}[for="${id}"]`) as HTMLCalciteLabelElement)) ||
+    closestElementCrossShadowBoundary(componentEl, labelTagName)
```

Comment:
I think we talked about this last time, but we won't expect nested labelable components, right?

Reply:
Correct.